internal/stores/objects_test.go

```diff
@@ -321,203 +318,175 @@ func TestSQLList(t *testing.T) {
+	// add the first two contracts to the autopilot
+	if err = db.SetContractSet("autopilot", []types.FileContractID{fcid1, fcid2}); err != nil {
+		t.Fatal(err)
 	}
 
-	// Create object.
+	// add an object
 	obj := object.Object{
 		Key: object.GenerateEncryptionKey(),
 		Slabs: []object.SlabSlice{
-			// 2/3 sectors bad
+			// good slab
 			{
 				Slab: object.Slab{
 					Key:       object.GenerateEncryptionKey(),
 					MinShards: 1,
 					Shards: []object.Sector{
-						sectorGood,
-						sectorBad,
-						sectorDeleted,
-					},
-				},
-			},
-			// 1/3 sectors bad
-			{
-				Slab: object.Slab{
-					Key:       object.GenerateEncryptionKey(),
-					MinShards: 1,
-					Shards: []object.Sector{
-						sectorGood,
-						sectorGood,
-						sectorDeleted,
+						{
+							Host: hk1,
+							Root: consensus.Hash256{1},
+						},
+						{
+							Host: hk2,
+							Root: consensus.Hash256{2},
+						},
+						{
+							Host: hk2,
+							Root: consensus.Hash256{3},
+						},
 					},
 				},
 			},
-			// 2/3 sectors bad
+			// unhealthy slab - hk3 is bad (1/3)
 			{
 				Slab: object.Slab{
 					Key:       object.GenerateEncryptionKey(),
 					MinShards: 1,
 					Shards: []object.Sector{
-						sectorBad,
-						sectorBad,
-						sectorGood,
+						{
+							Host: hk1,
+							Root: consensus.Hash256{4},
+						},
+						{
+							Host: hk2,
+							Root: consensus.Hash256{5},
+						},
+						{
+							Host: hk3,
+							Root: consensus.Hash256{6},
+						},
 					},
 				},
 			},
-			// 2/3 sectors bad
+			// unhealthy slab - hk3 is bad (2/3)
 			{
 				Slab: object.Slab{
 					Key:       object.GenerateEncryptionKey(),
 					MinShards: 1,
 					Shards: []object.Sector{
-						sectorBad,
-						sectorDeleted,
-						sectorGood,
+						{
+							Host: hk1,
+							Root: consensus.Hash256{7},
+						},
+						{
+							Host: hk3,
+							Root: consensus.Hash256{8},
+						},
+						{
+							Host: hk3,
+							Root: consensus.Hash256{9},
+						},
 					},
 				},
 			},
-			// 2/3 sectors bad
+			// unhealthy slab - hk4 is deleted (1/3)
 			{
 				Slab: object.Slab{
 					Key:       object.GenerateEncryptionKey(),
 					MinShards: 1,
 					Shards: []object.Sector{
-						sectorDeleted,
-						sectorDeleted,
-						sectorGood,
+						{
+							Host: hk1,
+							Root: consensus.Hash256{10},
+						},
+						{
+							Host: hk2,
+							Root: consensus.Hash256{11},
+						},
+						{
+							Host: consensus.PublicKey{4},
+							Root: consensus.Hash256{12},
+						},
 					},
 				},
 			},
-			// 3/3 sectors bad
+			// good slab - reused hosts
 			{
 				Slab: object.Slab{
 					Key:       object.GenerateEncryptionKey(),
 					MinShards: 1,
 					Shards: []object.Sector{
-						sectorBad,
-						sectorBad,
-						sectorDeleted,
+						{
+							Host: hk1,
+							Root: consensus.Hash256{1},
+						},
+						{
+							Host: hk1,
+							Root: consensus.Hash256{2},
+						},
+						{
+							Host: hk2,
+							Root: consensus.Hash256{3},
+						},
```

Comment:
I think we agreed once we would not keep track of sector roots on hosts we migrated away from. So what I'm saying is that the roots in a slab are unique and this edge case doesn't happen because we "drop" the "bad" sector from the slab if we migrate away from the host it's on

Reply:
Or do you mean cross objects/slabs?